### PR TITLE
8314021: HeapDump: Optimize segmented heap file merging phase

### PR DESCRIPTION
**Pull request by y1yang0**
This patch reduce ~16%(24s->20s) pahse 2 merge time during dumping 32g heap with 96threads and fixes a memory leak of compressor

You might argue why this is Linux-only optimization, because sendfile requires at least socket fd in other platforms([aix sendfile](https://www.ibm.com/docs/en/aix/7.1?topic=s-send-file-subroutine) [maxos sendfile](https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/sendfile.2.html) [win32 TransmitFile](https://learn.microsoft.com/en-us/windows/win32/api/mswsock/nf-mswsock-transmitfile)), while [only Linux](https://man7.org/linux/man-pages/man2/sendfile.2.html) supports both two file descriptors.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8314021](https://bugs.openjdk.org/browse/JDK-8314021): HeapDump: Optimize segmented heap file merging phase (**Enhancement** - P4)

### Reviewers
 * [Alex Menkov](https://openjdk.org/census#amenkov) (@alexmenkov - **Reviewer**)
 * [Kevin Walls](https://openjdk.org/census#kevinw) (@kevinjwalls - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15245/head:pull/15245` \
`$ git checkout pull/15245`

Update a local copy of the PR: \
`$ git checkout pull/15245` \
`$ git pull https://git.openjdk.org/jdk.git pull/15245/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15245`

View PR using the GUI difftool: \
`$ git pr show -t 15245`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15245.diff">https://git.openjdk.org/jdk/pull/15245.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15245#issuecomment-1674479235)

### CODE REVIEW

**Comment by y1yang0** [comment]
Gentle Ping :0 Can I have a review for this?

**Comment by y1yang0** [comment]
Can I have a review for this? This contains a memory leak and harmless optimziation.

File: src/hotspot/share/services/heapDumper.cpp
Comment (alexmenkov) [inline src/hotspot/share/services/heapDumper.cpp:1651]: Looks like you don't need separate method for linux.
Would it be better to make linux-specific implementation of merge_file():
#ifdef LINUX
// Merge segmented heap files via sendfile
void DumpMerger::merge_file(char* path) {
...
}
#else
// Generic implementation using read+write
void DumpMerger::merge_file(char* path) {
...
}
#endif

**Comment by alexmenkov** [comment]
The fix looks good to me in general, but I'm not sure about code organization.
src/hotspot/share/runtime/os.hpp describes rules for os* files.
It states:
//   Platform-independent source files should not include these header files
//   (although sadly there are some rare exceptions ...)
And the change adds one more exception.
I'd like to hear runtime guys opinion.

**Comment by y1yang0** [comment]
> The fix looks good to me in general, but I'm not sure about code organization. src/hotspot/share/runtime/os.hpp describes rules for os* files. It states: // Platform-independent source files should not include these header files // (although sadly there are some rare exceptions ...) And the change adds one more exception. I'd like to hear runtime guys opinion.

Thanks for the review, yes, grepping `os_linux.hpp` found several exceptions, I'd like to see if runtime forks have alternatives, too.

/label add hotspot-runtime

**Comment by y1yang0** [comment]
Hi @alexmenkov, can we move on since there are no more comments from runtime folks. We can still propose a follow-up change if someone have strong objection/concern on this "exception"

**Review by alexmenkov — APPROVED** [review]

**Comment by y1yang0** [comment]
Can I have a second review? Maybe @plummercj @kevinjwalls

**Review by kevinjwalls — APPROVED** [review]
Yes, looks good to me.
Apologies for delay, had looked at this previously but not come back for final review.

**Comment by y1yang0** [comment]
> Apologies for delay, had looked at this previously but not come back for final review.

Never mind:) Thanks @alexmenkov and @kevinjwalls for the reviews.

/integrate

**Comment by y1yang0** [comment]
> I had not noticed this PR sorry. I do not like the fact we have Linux-only functionality being added with no intent to supply similar functionality on other platforms. I also do not like the fact we had to ifdef the Linux code into the shared code. It would have been metter to place the code in os::Linux and then have a single LINUX_ONLY() to make that call.

Okay, I'll prepare a follow-up PR to do this.

**Comment by y1yang0** [comment]
> I had not noticed this PR sorry. I do not like the fact we have Linux-only functionality being added with no intent to supply similar functionality on other platforms. I also do not like the fact we had to ifdef the Linux code into the shared code. It would have been metter to place the code in os::Linux and then have a single LINUX_ONLY() to make that call.

I made a simple attempt, but it is difficult to integrate this portion of code into OS Linux because it requires modifying the writer's written bytes and setting error messages. Logically, it belongs to DumpMerge. Abstracting it into a function similar to concatenate_file and placing it into OS Linux seems challenging.

**Comment by dholmes-ora** [comment]
I think it can be improved somewhat even if we can't simply pass the `_writer` through.

**Pull request closed without merging**
